### ISSUE TITLE
Support HTML Table repr in Jupyter

### ISSUE THREAD

**Issue by onyxfish** [wireservice/agate#416]
Currently console `repr` prints the table's structure in the console and the table's contents (as HTML) in Jupyter. This should probably be consistent. It seems wrong to ever print the *entire* table as the result of a trivial `print(table)`. Possibilities:

1. Print the table's memory location. (Default repr behavior.)
2. Print the table's structure. (Current console behavior.)
3. Print the entire table data. (Current Jupyter behavior.)
4. Print a subset of the table. (Max x columns, max y rows.)

**Comment by onyxfish** [comment]
@ghing @TaurusOlson Would appreciate both of your input on this issue since you hacked on the relevant pull requests.

**Comment by onyxfish** [comment]
I suddenly remember that I had all of this working with repr at one point and it felt like the wrong behavior. I think the structure printing should happen when the table is stringified, rather than when it's repr'ed. (Which is what Taurus suggested.) The question then is whether it makes sense for Jupyter to behave differently, given that the HTML rendering hook only works for repr.

**Comment by onyxfish** [comment]
I've reverted `repr` to display memory location for console usage and have implemented `__str__` to print the table structure. This functionality seems correct to me.

I'm not sure the right behavior for Jupyter. On the one hand, Jupyter encourages overriding `repr` to display partial contents ([pandas displays a subset of rows and columns](http://pandas.pydata.org/pandas-docs/stable/options.html)). On the other hand, this would just be an arbitrary choice without implementing the same sort of options that pandas has. (Since you can't pass a parameter to `repr`.)

Also, Jupyter's usage of repr is [not what repr is for](https://docs.python.org/2/library/functions.html#func-repr), so that's a bit frustrating.

**Comment by TaurusOlson** [comment]
Hi @onyxfish,
I haven't thought about that for a while. I'll get the latest version of agate and will look at that tomorrow or the day after. 

> I think the structure printing should happen when the table is stringified, rather than when it's repr'ed. (Which is what Taurus suggested.)

My initial proposition was to use the method `print_structure` and you thought the result would make a good output for `__unicode__`. I don't really know the difference between `__unicode__` and `__str__` but I think you made the right choice: `repr` should display memory location for console usage and `__str__` should print the table structure.

I'll play with agate in the Jupyter notebook and will let you know what I think.

**Comment by onyxfish** [comment]
Hey @TaurusOlson, thanks for the feedback. FYI, `__str__` is simply the Python 3 name for `__unicode__`. The way I implemented it will work for Python 2 as well.

**Comment by ghing** [comment]
The thing that sucks is that Jupyter conventions and Python conventions seem at odds here.  It's really helpful to be able to see the table quickly with `repr` in Jupyter, but I'd be open to a compromise where the ability to output the table as HTML is available as a public method (which I believe @onyxfish already implemented) of `agate.Table` and the `repr` behavior exists in a plugin.

**Comment by onyxfish** [comment]
@ghing I think that's an good compromise. The `Table.print_html` method can be used to implement this in a plugin with very little overhead. In fact, that's how it's implemented now:

```python
    def _repr_html_(self):
        """
        Print HTML version of the table for Jupyter
        (via :meth:`Table.print_html`).

        See http://ipython.readthedocs.org/en/stable/config/integrating.html?highlight=_repr_html_#rich-display
        """
        html = six.StringIO()

        self.print_html(output=html)

        return html.getvalue()
```
 I'm going to remove the `_repr_html_` method from core, for now, but I'm open to bringing it back into core in the future if we can figure out an elegant way.

It seems the ideal solution would be if there was a way to flag for Jupyter that a given string is HTML. (With a subclass of the string object, for instance.) Then we could more easily support this case.

**Comment by TaurusOlson** [comment]
I gave a look to the latest version (commit `b891191`) and everything we talked about looks good. However, in Jupyter `print_html` returns the string of html table not the html table I expected. Is this normal?

**Comment by onyxfish** [comment]
Yes, there is nothing directly I can do about that. It'll only render the HTML if it's returned from the `_repr_html_` method. See above for my thinking around removing that from this release. Ideally, Jupyter would provide some other flag to render that correctly.

**Comment by onyxfish** [comment]
I've wasted hours trying to resolve this on at least three separate occasions. It can't be solved to my own satisfaction. The inconsistency between the behavior of repr and str annoys me too much to fold it into core, especially given what a mess it makes of the code. Also, at the end of the day, I'm not sure the formatted HTML tables are actually even more readable than the plain-text ones.

Note to future me: next time work on something more useful instead. You aren't going to find a solution to this that you like.

Closing.
